osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -196,16 +198,46 @@ protected override void UpdateAfterChildren()
 
                 float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
                 Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
+                Func<T, float> sectionPos = section => scrollContainer.GetChildPosInContent(section);
+                Func<T, float> visibleSize = section => Math.Min(
+                    sectionPos(section) + section.Height, currentScroll + scrollContainer.DisplayableContent) - Math.Max(sectionPos(section), currentScroll
+                );
+                var sortedByVisible = new List<T>(Children);
+                sortedByVisible.Sort((a, b) => visibleSize(b).CompareTo(visibleSize(a)));
+                T mostVisible = sortedByVisible.FirstOrDefault();
+
+                if (mostVisible != null)
+                {
+                    int mostVisibleIndex = IndexOf(mostVisible);
+                    T ch;
+
+                    while (mostVisibleIndex != 0 && Precision.AlmostEquals(visibleSize(ch = Children[mostVisibleIndex - 1]), ch.Height))
+                    {
+                        mostVisibleIndex--;
+                    }
+
+                    SelectedSection.Value = Children[mostVisibleIndex];
+
+                    if (Precision.AlmostBigger(scrollOffset, currentScroll))
+                    {
+                        SelectedSection.Value = Children.FirstOrDefault();
+                    }
 
-                if (scrollContainer.IsScrolledToEnd())
+                    if (Precision.AlmostBigger(currentScroll, scrollContainer.ScrollableExtent))
+                    {
+                        SelectedSection.Value = Children.LastOrDefault();
+                    }
+                }
+
+                /*if (scrollContainer.IsScrolledToEnd())
                 {
                     SelectedSection.Value = Children.LastOrDefault();
                 }
                 else
                 {
                     SelectedSection.Value = Children.TakeWhile(section => diff(section) <= 0).LastOrDefault()
                                             ?? Children.FirstOrDefault();
-                }
+                }*/
```

Comment:
Opening as draft is no excuse to leave behind dead commented code.

Reply:
I thought this the easiest way to test master vs this PR. Also the reason why ``diff`` is still there and caused CI failure.

Reply:
Dynamic recompilation works with `git checkout` as far as I know, so actually the easiest way to check would be to detached-head-checkout d957679 and then ac94b2d.